File: app/src/main/java/com/yu/functionbox/function/FunctionActivity.java
package com.yu.functionbox.function;

import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;

import com.yu.functionbox.R;
import com.yu.functionbox.databinding.ActivityFunctionBinding;

public class FunctionActivity extends AppCompatActivity {
    private ActivityFunctionBinding mBinding;
    private FunctionViewModel mViewModel;
    public static final String FLAG_CREATE_NEW = "flag_create_new";
    public static final String FLAG_CHECK = "flag_check";
    public static final String KEY_CONTENT = "key_content";
    public static final String KEY_ID = "key_id";
    public static final String KEY_TYPE = "key_type";
    public static final int RESULT_FUNCTION = 101;
    public final static int REQUEST_FUNCTION = 1001;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBinding = DataBindingUtil.setContentView(this,R.layout.activity_function);
        mViewModel = new FunctionViewModel(this);
        mBinding.setViewModel(mViewModel);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        if(getSupportActionBar()!=null){
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
        toolbar.setNavigationOnClickListener(v -> finish());

        initData();
    }

    private void initData() {
        Intent intent = getIntent();
        switch (intent.getStringExtra(KEY_TYPE)){
            case FLAG_CREATE_NEW:
                mViewModel.enableEdit(true);
                break;
            case FLAG_CHECK:
                String content = intent.getStringExtra(KEY_CONTENT);
                long id = intent.getLongExtra(KEY_ID,0L);
                mViewModel.setContent(id,content);
                break;
        }
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        finish();
    }

    @Override
    public void finish() {
        if(mViewModel.isEdited()){
            this.setResult(RESULT_FUNCTION);
        }
        super.finish();
    }
}

File: app/src/main/java/com/yu/functionbox/function/FunctionActivity.kt
package com.yu.functionbox.function

import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import com.yu.functionbox.R
import com.yu.functionbox.databinding.ActivityFunctionBinding

class FunctionActivity : AppCompatActivity() {
    private lateinit var mBinding: ActivityFunctionBinding
    private lateinit var mViewModel: FunctionViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_function)
        mViewModel = FunctionViewModel(this)
        mBinding.viewModel = mViewModel
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        if (supportActionBar != null) {
            supportActionBar?.setDisplayHomeAsUpEnabled(true)
        }
        toolbar.setNavigationOnClickListener { finish() }

        initData()
    }

    private fun initData() {
        val intent = intent
        when (intent.getStringExtra(KEY_TYPE)) {
            FLAG_CREATE_NEW -> mViewModel.enableEdit(true)
            FLAG_CHECK -> {
                val content = intent.getStringExtra(KEY_CONTENT)
                val id = intent.getLongExtra(KEY_ID, 0L)
                mViewModel.setContent(id, content)
            }
        }
    }

    override fun onBackPressed() {
        super.onBackPressed()
        finish()
    }

    override fun finish() {
        if (mViewModel.isEdited) {
            this.setResult(RESULT_FUNCTION)
        }
        super.finish()
    }

    companion object {
        val FLAG_CREATE_NEW = "flag_create_new"
        val FLAG_CHECK = "flag_check"
        val KEY_CONTENT = "key_content"
        val KEY_ID = "key_id"
        val KEY_TYPE = "key_type"
        val RESULT_FUNCTION = 101
        val REQUEST_FUNCTION = 1001
    }
}
